fedimint-client/src/module/mod.rs

```diff
@@ -725,38 +645,69 @@ pub trait IClientModule: Debug {
 
     async fn backup(&self, module_instance_id: ModuleInstanceId)
         -> anyhow::Result<DynModuleBackup>;
+}
 
-    fn supports_being_primary(&self) -> bool;
-
+#[apply(async_trait_maybe_send!)]
+pub trait BalanceProvider: 'static + MaybeSend + MaybeSync {
```

Comment:
gist of PR: move methods of primary module to a trait that can be implemented by anything, not just a module

Reply:
I think directly implementing this for modules will be challenging. Primary modules had a separate trait in the beginning (much like proposed here, just with a different name) and it was a huge PITA. 

What could work is a mix of both approaches: the module trait has a `fn as_balance_provider(&self) -> Option<DynBalanceProvider>` which can return an independent struct or a clone of the module, just put into a different dyn box.

When specifying a balance provider we have to keep in mind that modules aren't initialized yet when we call builder functions. I'd propose giving the builder a function that, given the set of all initialized modules, creates a `DynBalanceProvider`. This function could either select one module or construct a more complicated struct that contains multiple balance providers and selects the appropriate one depending on context. This fn could default to selecting the first module returning `Some`.

Reply:
`Module` don't implement `BalanceProvider`, `ModuleBalanceProvider`.
trick is to make `BalanceProvider` state free, it takes a `Client` as argument in all functions. 

we can do
```
builder.with_balance_provider(mint_client::MintBalanceProvider::default())
```

Reply:
What's the benefit of moving to another trait? Balance provider could work with `DynClientModule` all the same, and filter out ones that don't support it.

It seems really like an aestetic benefit only. @maan2003

Reply:
Thx for correcting my understanding @maan2003. I tend to agree with @dpc now (I think). 

The module trait can stay the same and the `BalanceProvider` trait can be seen as a tuple of two closures that, given all client modules, will create an input/output respectively (just like primary modules do).

A `DefaultBalanceProvider` could e.g. scan through all modules, looking for the first primary one and using that. If we ever had multiple e-cash modules and wanted to migrate from one to the other we'd make it more sophisticated and use all modules for creating inputs (with priority on the old one) and only create outputs for the new one, so over time all funds are migrated.

I don't know how to achieve that yet, but I could also imagine using different balance providers for different transactions eventually:

```rust
client.get_first_module::<LightningClientModule>()
    .with_balance_provider(MyBalanceProvider)
    .pay_bolt11_invoice(…)
    .await?;
```

Reply:
oh per transaction is interesting

Reply:
we can just make `pay_bolt11_invoice` have a balance provider argument, call site is always more flexible than storing  stuff